examples/tls-sni/envoy.yaml

```diff
@@ -0,0 +1,150 @@
+              domains:
+              - "*"
+              routes:
+              - match:
+                  prefix: "/"
+                route:
+                  cluster: proxy-domain1
+          http_filters:
+          - name: envoy.filters.http.router
+      transport_socket:
+        name: envoy.transport_sockets.tls
+        typed_config:
+          "@type": type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.DownstreamTlsContext
+          common_tls_context:
+            tls_certificates:
+              certificate_chain:
+                filename: certs/domain1.crt.pem
+              private_key:
+                filename: certs/domain1.key.pem
+
+    - filter_chain_match:
+        server_names:
+        - domain2.example.com
+      filters:
+      - name: envoy.filters.network.http_connection_manager
+        typed_config:
+          "@type": type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager
+          codec_type: auto
+          stat_prefix: ingress_http
+          route_config:
+            name: local_route
+            virtual_hosts:
+            - name: app
+              domains:
+              - "*"
+              routes:
+              - match:
+                  prefix: "/"
+                route:
+                  cluster: proxy-domain2
+          http_filters:
+          - name: envoy.filters.http.router
+      transport_socket:
+        name: envoy.transport_sockets.tls
+        typed_config:
+          "@type": type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.DownstreamTlsContext
+          common_tls_context:
+            tls_certificates:
+              certificate_chain:
+                filename: certs/domain2.crt.pem
+              private_key:
+                filename: certs/domain2.key.pem
+
+    - filter_chain_match:
+        server_names:
+        - domain3.example.com
+      filters:
+      - name: envoy.filters.network.http_connection_manager
+        typed_config:
+          "@type": type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager
+          codec_type: auto
+          stat_prefix: ingress_http
+          route_config:
+            name: local_route
+            virtual_hosts:
+            - name: app
+              domains:
+              - "*"
+              routes:
+              - match:
+                  prefix: "/"
+                route:
+                  cluster: proxy-domain3
+          http_filters:
+          - name: envoy.filters.http.router
+      transport_socket:
+        name: envoy.transport_sockets.tls
+        typed_config:
+          "@type": type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.DownstreamTlsContext
+          common_tls_context:
+            tls_certificates:
+              certificate_chain:
+                filename: certs/domain3.crt.pem
+              private_key:
+                filename: certs/domain3.key.pem
```

Comment:
I think it would be useful if one of those filter chains didn't do TLS terminator in Envoy and instead do SNI routing, with TLS terminated by the backed.

Reply:
oh ok - i reckon thats pretty doable

i can update tomorrow - @mattklein123 perhaps /wait while i update

Reply:
SNI is detected by the TLS inspector by peeking at the first packet (TLS ClientHello, which contains the SNI in cleartext). This results in end-to-end TLS between client and backend, and not 2 TLS sessions, one between client and proxy, and another between proxy and backend.

This can be achieved by removing `transport_socket` from the last filter chain, and adding TLS configuration to the appropriate backend.

Reply:
i removed the `transport_socket` for the last listener and added a tls endpoint for the third upstream - but i dont seem to be able to make it work

im struggling to understand how it would work if it still has the http_connection_manager

are you suggesting just tcp passthrough ?

or does the upstream endpoint have to have sni enabled for this to work ?

can you suggest a specific config for the third listener that would work as you suggest

Reply:
Ah, yeah, sorry... it should be TCP proxy, not HTTP connection manager.

Endpoint doesn't have to have the SNI enabled, since SNI routing is done in Envoy.

Reply:
cool. i got it working.

i need to update the docs a bit to reflect the changes